common/oauthTokenManager.go

```diff
@@ -682,59 +641,48 @@ func (credInfo *OAuthTokenInfo) GetWorkloadIdentityCredential() (azcore.TokenCre
 	return tc, nil
 }
 
-type DeviceCodeCredential struct {
-	token       adal.Token
-	aadEndpoint string
-	tenantID    string
-	clientID    string
-}
-
-func (dcc *DeviceCodeCredential) GetToken(ctx context.Context, options policy.TokenRequestOptions) (azcore.AccessToken, error) {
-	waitDuration := dcc.token.Expires().Sub(time.Now().UTC()) / 2
-	if dcc.token.WillExpireIn(waitDuration) {
-		resource := strings.TrimSuffix(options.Scopes[0], "/.default")
-		_, err := dcc.RefreshTokenWithUserCredential(ctx, resource)
-		if err != nil {
-			return azcore.AccessToken{}, err
-		}
-	}
-	return azcore.AccessToken{Token: dcc.token.AccessToken, ExpiresOn: dcc.token.Expires()}, nil
-}
-
-// RefreshTokenWithUserCredential gets new token with user credential through refresh.
-func (dcc *DeviceCodeCredential) RefreshTokenWithUserCredential(ctx context.Context, resource string) (*adal.Token, error) {
-	targetResource := resource
-	if dcc.token.Resource != "" && dcc.token.Resource != targetResource {
-		targetResource = dcc.token.Resource
-	}
-
-	oauthConfig, err := adal.NewOAuthConfig(dcc.aadEndpoint, dcc.tenantID)
+func (credInfo *OAuthTokenInfo) GetDeviceCodeCredential() (azcore.TokenCredential, error) {
+	authorityHost, err := getAuthorityURL(credInfo.Tenant, credInfo.ActiveDirectoryEndpoint)
 	if err != nil {
 		return nil, err
 	}
-
-	// ClientID in credInfo is optional which is used for internal integration only.
-	// Use AzCopy's 1st party applicationID for refresh by default.
-	spt, err := adal.NewServicePrincipalTokenFromManualToken(
-		*oauthConfig,
-		Iff(dcc.clientID != "", dcc.clientID, ApplicationID),
-		targetResource,
-		dcc.token)
+	var persistenceOptions *azidentity.TokenCachePersistenceOptions
+	if credInfo.Persist {
+		persistenceOptions = &azidentity.TokenCachePersistenceOptions{
+			Name: TokenCache,
+		}
+	}
+	// Read the record=
+	record := IffNotNil(credInfo.DeviceCodeInfo, azidentity.AuthenticationRecord{})
+	tc, err := azidentity.NewDeviceCodeCredential(&azidentity.DeviceCodeCredentialOptions{
+		TenantID:                       credInfo.Tenant,
+		ClientID:                       ApplicationID,
+		DisableAutomaticAuthentication: true,
+		TokenCachePersistenceOptions:   persistenceOptions,
+		AuthenticationRecord:           record,
+		ClientOptions: azcore.ClientOptions{
+			Cloud:     cloud.Configuration{ActiveDirectoryAuthorityHost: authorityHost.String()},
+			Transport: newAzcopyHTTPClient(),
+		},
+	})
 	if err != nil {
 		return nil, err
 	}
 
-	if err := spt.RefreshWithContext(ctx); err != nil {
-		return nil, err
+	if record == (azidentity.AuthenticationRecord{}) {
+		// No stored record; call Authenticate to acquire one
+		record, err = tc.Authenticate(context.TODO(), &policy.TokenRequestOptions{Scopes: []string{StorageScope}})
+		if err != nil {
+			return nil, err
+		}
+		if credInfo.Tenant == DefaultTenantID {
```

Comment:
Could you please help me explain a scenario why this needs to be sync.Once? I can't see tenant as a shared resource in the login workflow. Please correct me if I am wrong.

Reply:
On the off-chance we call GetDeviceCodeCredential multiple times for whatever reason, stdout might look like...

```
INFO: Logging in under the...
INFO: If you plan to use...
INFO: Logging in under the...
INFO: If you plan to use...
INFO: Logging in under the...
INFO: If you plan to use...
```

This is quite spammy. Granted, it would be a decent indicator a mistake has been made, though.